### ISSUE TITLE
Composite Id

### ISSUE THREAD

**Issue by umk8yw** [javers/javers#625]
I would be glad to contribute at implementation of a solution for defining multiple fields as an identifier for a business model.

**Comment by bartoszwalacik** [comment]
Good to hear! Can we start from writing a failing test with some use cases? I can give you access to javers repo to make our work easier.

**Comment by umk8yw** [comment]
I was thinking to change to `EntityType` with a `Set<JaversProperty>` instead of the` JaversProperty idProperty` field and adapt the rest of the functionality.

**Comment by bartoszwalacik** [comment]
bare Set is not ok in terms of OO modeling but we can create a dedicated class fot that, lets say EntityId and hide this set

**Comment by umk8yw** [comment]
ok, this was my approach into `EntityDefinitionBuilder`, I've created a dedicated class to store `Set<String> propertyNames` and I've created method: 
```
public EntityDefinitionBuilder withCompositeId(String... compositeIdPropertyNames) {
    	Validate.argumentsAreNotNull(compositeIdPropertyNames);
    	this.uniqueKey = Optional.of(new UniqueKey(compositeIdPropertyNames));
    	return this;
    }
```
so when Javers is built to have the possibility to add multiple fields as id

**Comment by bartoszwalacik** [comment]
OK, now when you have access to javers repo, could you come up with a failing E2E test and push it to a new branch? This test should be a specification for our new Composite Id feature.

**Comment by umk8yw** [comment]
when I tried to run again the test I found another problem at some code added on 8 December:

```
 private String localIdAsString(Object localId) {
        PrimitiveOrValueType idPropertyType = getIdProperty().getType();
        return idPropertyType.smartToString(localId);
    }
```

because my id is a `ComplexEntity` this method fails.
I've done a pull request so you can see exactly what I am talking about, its title is _Composite Id #625 - found problem with new method "localIdAsString"_

**Comment by bartoszwalacik** [comment]
so, javers clashes because you have forced JaversType of your Id Property to EntityType:

`.registerEntity(entityDefinition(ComplexEntity.class).withIdPropertyName("code").build())`
`registerEntity(entityDefinition(TenderLocation.class).withIdPropertyName("complexEntity").build()`

It's not possible for now, JaversType of an Id Property can be only Primitive or Value.

**Comment by bartoszwalacik** [comment]
btw, i'm not sure why this kind of example,
I thought that we want to achieve sth like this:

```
class Person {
    @Id String name
    @Id String surname
    @Id LocalDate dob
    ...
}

**Comment by umk8yw** [comment]
yes, I want to be able to define multiple fields as an id, but I want to be able to do this at `javers` build, and some of the fields defined as `id` will not be Primitive or Value.

in the pull request is exactly the example of the way I've imagined it.

**Comment by bartoszwalacik** [comment]
You are talking about 2 different issues, one issue is composite Id, so possibility to point more than one Id property for an Entity (with both ways, annotations and javersBuilder). That's what we agreed on.
Second issue is possibility to use Entity as Id property, and that's what I don't want in javers because it brakes the rule that Id property can be only Primitive or Value (see https://javers.org/documentation/domain-configuration/#entity-id). I don't want to change this rule, moreover, I think that this is not a good idea in terms of domain model. Why to put one Entity as an Id of another Entity?

**Comment by umk8yw** [comment]
ok, then what solution do you see for this kind of situation? 
I've chosen this approach because I have 3 entities which all have the `equals `method overriden and I have to specify their id because all of then have multiple fields defined as id. 
more, there is an hierarchical relation between entities.
ComplexEntity
    |__ List<Location>
             |__ List<Airline>
and `Location` has in its unique key `ComplexEntity` and `Airline` also has in its unique key a `Location`.

my first try was to make all of them `ValueObject` but the result was not the one expected.

**Comment by bartoszwalacik** [comment]
@umk8yw i don't see Airline and Location classes, could you express your issue in more general way? What feature or features you want to add to JaVers finally? Would this features be useful for other Javers users?

**Comment by ismaelgomescosta** [comment]
@bartoszwalacik I opened a solution for this issue. and its a simple change needed to make this case possible.

https://github.com/javers/javers/pull/644

In my case I have an entity which is the Id of another entity. 

Like 
Person
    |___  PersonDocumentation       
Where

Documentation
   |___  
       @id  Person

I can't change this structure because Person saves personDocumentation by cascade and the Id of my documentation is the Id of my People. And i can`t change this because I work in a legacy project.

Please help me with that.

**Comment by bartoszwalacik** [comment]
This issue is about making this mapping possible

```
class Person {
    @Id String name
    @Id String surname
    @Id LocalDate dob
    ...
}
```
usually, this is called composite Id

**Comment by singhjoga** [comment]
Hi,
I wanted to use Javers in my project where many entities have multiple Id fields and entity itself is annotated with IdClass.
Is there any further progress on this issue? If not, maybe with your guidance, I can try to implement it.

Best Regards,

Joga

**Comment by bartoszwalacik** [comment]
@singhjoga nobody is working on this issue now, your contribution is welcome. I can guide you by writing first failing test case.

**Comment by singhjoga** [comment]
Hi @bartoszwalacik,
Yes please, send me the information to get started. I have already forked the git repo from master.

Joga

**Comment by bartoszwalacik** [comment]
OK, so you can start from adding support for composite Id to the diff algorithm (`javers.compare()`) and then to the persistence layer (`javers.commit()`). New tests for comparing should be placed close to `JaversDiffE2ETest`. 

In your test, write some Class, for example:

```
Person class {
    @Id private String firstName;
    @Id private String lastName; 
}
```
and try to compare some objects.

**Comment by shiruba** [comment]
I would like to contribute. I already added tests to JaversCommitE2ETest and JaversDiffE2ETest, which are failing as expected. Now I am trying to understand what I should do to fix these tests. 

My guess is that I need to change the `EntityType` so that it can hold more than one idProperty. This also involves the `EntityTypeFactory`, which creates the `EntityType`.

What should I do after that?

**Comment by bartoszwalacik** [comment]
Then `GlobalIdFactory`, the place where `InstanceId` is created.
`InstanceId` holds object's Id as Object. If we don't want to change this structure, we can put `List` of all values grabbed from fields marked as `@Id`

```
 private final Object cdoId;
```

**Comment by bartoszwalacik** [comment]
released in 5.4.0, see https://javers.org/release-notes
